lib/manifold.ex

```diff
@@ -1,42 +1,63 @@
 defmodule Manifold do
   use Application
 
-  alias Manifold.{Partitioner, Utils}
+  alias Manifold.Partitioner
+  alias Manifold.Sender
+  alias Manifold.Utils
 
   @max_partitioners 32
   @partitioners min(Application.get_env(:manifold, :partitioners, 1), @max_partitioners)
   @workers_per_partitioner Application.get_env(:manifold, :workers_per_partitioner, System.schedulers_online)
 
+  @max_senders 128
+  @senders min(Application.get_env(:manifold, :senders, 1), @max_senders)
+
   ## OTP
 
   def start(_type, _args) do
     import Supervisor.Spec, warn: false
 
-    children = for partitioner_id <- 0..(@partitioners - 1) do
+    partitioners = for partitioner_id <- 0..(@partitioners - 1) do
       Partitioner.child_spec(@workers_per_partitioner, [name: partitioner_for(partitioner_id)])
     end
+    senders = for sender_id <- 0..(@senders - 1) do
+      Sender.child_spec([name: sender_for(sender_id)])
+    end
 
-    Supervisor.start_link children,
+    Supervisor.start_link partitioners ++ senders,
       strategy: :one_for_one,
       max_restarts: 10,
       name: __MODULE__.Supervisor
   end
 
   ## Client
 
-  @spec send([pid | nil] | pid | nil, term) :: :ok
-  def send([pid], message), do: __MODULE__.send(pid, message)
-  def send(pids, message) when is_list(pids) do
-    partitioner_name = current_partitioner()
-    grouped_by = Utils.group_by(pids, fn
-      nil -> nil
-      pid -> node(pid)
-    end)
-    for {node, pids} <- grouped_by, node != nil, do: Partitioner.send({partitioner_name, node}, pids, message)
-    :ok
+  @spec send([pid | nil] | pid | nil, term, send_mode: :offload) :: :ok
+  def send(pid, message, options \\ [])
+  def send([pid], message, options), do: __MODULE__.send(pid, message, options)
+  def send(pids, message, options) when is_list(pids) do
+    case options[:send_mode] do
+      :offload ->
+        Sender.send(current_sender(), current_partitioner(), pids, message)
+      nil ->
+        partitioner_name = current_partitioner()
+        grouped_by = Utils.group_by(pids, fn
+          nil -> nil
+          pid -> node(pid)
+        end)
+        for {node, pids} <- grouped_by, node != nil, do: Partitioner.send({partitioner_name, node}, pids, message)
+        :ok
+    end
+  end
+  def send(pid, message, options) when is_pid(pid) do
+    case options[:send_mode] do
```

Comment:
I think this code path and the clause on line 37 are trying to make a carveout for sending to a single pid to make that efficient too.

Here the `:offload` flag doesn't really help us very much, since we are only sending to a single destination we end up paying the same penalty whether we send it directly to the remote partitioner or to the sender.  

I think it would make sense to not honor the flag here, it relieves the sender of some work that doesn't really help anyone and doesn't hurt the caller any.

Reply:
That's a good observation. I'll make the change.

Reply:
@ihumanable I think this optimization might not be safe. It could potentially break the linearizability guaranteed by `send/2`.

Consider this scenario with these two back-to-back calls:

```
Manifold.send([pid_1, pid_2], :message_1, send_mode: :offload)
Manifold.send(pid_1, :message_2, send_mode: :offload)
```

Wouldn't in theory `:message_2` could get sent by the caller process *before* `:message_1` gets sent by one of the `Manifold.Sender` processes?

Reply:
Yea, that does seem right.  Although this raises an even more interesting sharp edge and one we should document.

The linearization guarantee goes away if you mix calls to Manifold with and without offloading.  Someone might have some small messages where they want to skip offloading, so they sometimes call Manifold send for the same destinations with and without offloading.

We should make sure to clearly document that that too will cause linearization guarantees to fail.

Reply:
Good point. We are on the same page.